ganttproject/src/main/java/net/sourceforge/ganttproject/GanttProject.java

```diff
@@ -956,33 +951,62 @@ public void refresh() {
     super.repaint();
   }
 
-  // TODO: Accumulate changes instead of sending it every time.
+  private interface TxnSendListener {
+    void onSendCompleted();
+  }
+
+  private final AtomicReference<TxnSendListener> txnSendingListener = new AtomicReference<>();
+
   private Unit sendProjectStateLogs() {
     gpLogger.debug("Sending project state logs");
+    if (txnSendingListener.get() != null) return Unit.INSTANCE;
     try {
       var baseTxnCommitInfo = myBaseTxnCommitInfo.get();
       var txns = myProjectDatabase.fetchTransactions(baseTxnCommitInfo.right + 1, 1);
       if (!txns.isEmpty()) {
-        getWebSocket().sendLogs(new InputXlog(
-          baseTxnCommitInfo.left,
-          "userId",
-          "refid",
-          txns
-        ));
+        var listener = new TxnSendListener() {
+          @Override
+          public void onSendCompleted() {
+            txnSendingListener.compareAndSet(this, null);
+          }
+        };
+        if (txnSendingListener.compareAndSet(null, listener)) {
+          getWebSocket().sendLogs(new InputXlog(
+            baseTxnCommitInfo.left,
+            "userId",
+            "refid",
+            txns
+          ));
+        }
       }
     } catch (ProjectDatabaseException e) {
       gpLogger.error("Failed to send logs", new Object[]{}, ImmutableMap.of(), e);
     }
     return Unit.INSTANCE;
   }
 
+  @Override
+  protected Unit onProjectLogUpdate() {
+    if (isColloboqueLocalTest()) {
```

Comment:
I don't like that we have more and more these checks. I thought that we would have just a few differences between the "local test" and "prod" modes (url, authentication), but now it seems that we will just not do anything in prod.

Reply:
remove this check? And maybe the super call too?

Reply:
Hmmm
We perform websocket handshake (with base txn id & project refid exchange) only under this check. 
If we remove the check, we'll log an error every time a new log record is added for an offline document, won't we?

Reply:
Well, in this case we shall check "if we are working with an online document, in any sense", not "is it a local online document setup". 

We probably need to modify the local dev server so that it could respond to project read and write operations, just like a regular GP Cloud server. In this case the check may look like `if (project.document is OnlineDocument)`